### PR TITLE
CDAP-1756 allow templates to add datasets and streams when

### PR DESCRIPTION
**Pull request by albertshau**
configuring an adapter. Some refactoring along the way to make
sure that most of the code is shared between adapters and apps.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT1358

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationManager.java [hunk 1]
```diff
@@ -100,19 +99,19 @@ public LocalApplicationManager(CConfiguration configuration, PipelineFactory pip
...
+    pipeline.addLast(new DeployDatasetModulesStage(configuration, namespace, datasetFramework,
+                                                   inMemoryDatasetFramework));
+    pipeline.addLast(new CreateDatasetInstancesStage(configuration, datasetFramework, namespace));
+    pipeline.addLast(new CreateStreamsStage(namespace, streamAdmin, exploreFacade, exploreEnabled));
     pipeline.addLast(new DeletedProgramHandlerStage(store, programTerminator, streamConsumerFactory,
                                                     queueAdmin, discoveryServiceClient));
     pipeline.addLast(new ProgramGenerationStage(configuration, namespacedLocationFactory));
     pipeline.addLast(new ApplicationRegistrationStage(store));
-    pipeline.addLast(new CreateSchedulesStage(scheduler));
-    pipeline.setFinally(new DeployCleanupStage());
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationManager.java]: Why was this removed?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterServiceTests.java [hunk 2]
```diff
@@ -144,6 +147,18 @@ public void testRedeploy() throws Exception {
...
+    String streamName = "somestream";
+    String tableName = "sometable";
+    String adapterName = "streamAdapter";
+    DataTemplate.Config config = new DataTemplate.Config(streamName, tableName);
+    AdapterConfig adapterConfig = new AdapterConfig("description", DataTemplate.NAME, GSON.toJsonTree(config));
+    adapterService.createAdapter(NAMESPACE, adapterName, adapterConfig);
+
+    Assert.assertTrue(streamExists(Id.Stream.from(NAMESPACE, streamName)));
+    Assert.assertTrue(datasetExists(Id.DatasetInstance.from(NAMESPACE, tableName)));
+  }
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterServiceTests.java]: Need a line break

File: cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  @Override
+  public void configure() {
+    setName(NAME);
+    addWorker(new DWorker());
+  }
+
+  @Override
+  public void configureAdapter(String adapterName, Config config, AdapterConfigurer configurer) throws Exception {
+    configurer.addDatasetModule("customDataset", CustomDatasetModule.class);
+    if (config.streamName != null && !config.streamName.isEmpty()) {
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: So the configureAdapter method will decide what streams, datasets to be created from the Adapter configuration? How will configureAdapter know what streams, datasets to create? For example, a specific Source or Sink will know what stream, dataset it needs right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationManager.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationManager.java]: this was performing cleanup at the end. I changed it so the stage performs clean up on itself so that it is simpler and less error prone. For example, before, deploy dataset modules could unpack some jars, then you could have have an error in a later stage and those jars might not get cleaned up.

File: cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: yes, templates will have to declare them during configureAdapter.  For the etl case, this means the driver needs to allow the sources and sinks to add streams and datasets, which seems correct to me. You know everything that is needed to create required datasets at adapter creation time, and you can make sure it happens before any processing begins.

Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: Ok. Sounds good. It would be good to see how the source, sink APIs need to change to support this.

Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: yeah that is still up in the air. One way would be to add a configureAdapter() method to them that configureAdapter then calls.  Another way would be to pass in a config to the source/sink configure method, and add add stream/dataset methods to the StageConfigurer, which the etl driver uses to delegate to the AdapterConfigurer's methods.

Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: i would probably add a method like:

```configurePipeline(ETLStage config, PipelineConfigurer pipelineConfigurer)```

Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/DataTemplate.java]: The current configure method in Source/Sink is primarily used to convey the information about the properties that specific source/sink is expecting the user to provide. So it could/should be invoked before we have a adapter configuration.

**Comment by gokulavasan** [comment]
Comment about Source/Sink API change. We can figure that out in a separate PR. Rest LGTM :+1:
